### PR TITLE
GAWB-3194: API to support nowhere to save

### PR DESCRIPTION
**Pull request by davidangb**
New convenience API that returns two booleans:
* does the user have write access to at least one workspace?
* does the user have a ready billing project, which would allow her to create (and therefore write to) a workspace?

The UI can use this convenience API when checking permissions on Dockstore workflow imports. Otherwise, the UI would have to make two ajax requests to the billing project list and workspace list, then perform filtering on those results. The workspace list is byte-size O(N), but this convenience API will always be the same small size.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've followed [the instructions](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/FireCloudRequestBuilding.scala [hunk 1]
```diff
@@ -35,10 +35,10 @@ trait FireCloudRequestBuilding extends spray.httpx.RequestBuilding {
     addCredentials(OAuth2BearerToken("mF_9.B5f-4.1JqM"))
   }
 
-  def dummyUserIdHeaders(userId: String) = {
+  def dummyUserIdHeaders(userId: String, token: String = "access_token") = {
     addCredentials(OAuth2BearerToken("mF_9.B5f-4.1JqM")) ~>
       addHeader(RawHeader("OIDC_CLAIM_user_id", userId)) ~>
-      addHeader(RawHeader("OIDC_access_token", "access_token")) ~>
+      addHeader(RawHeader("OIDC_access_token", token)) ~>
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/FireCloudRequestBuilding.scala:41]: refactored this to allow unit tests to pass an arbitrary token value into their DAOs.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/UserService.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      hasWorkspace <- workspaces.map { ws => ws.exists(_.accessLevel.compare(WorkspaceAccessLevels.Write) >= 0)} // TODO: also need to check canCompute?
+    } yield UserImportPermission(
+      billingProject = hasProject,
+      writableWorkspace = hasWorkspace)
+
+    // respond with a UserImportPermission
+    importPerm map { x=> RequestComplete(StatusCodes.OK, x)}
+  }
+
+}
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/UserService.scala]: the TODO above ("need to check canCompute?") is an open question to PO, etc. in an email. I'll update the PR or make a new PR if we need to check that.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/UserService.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    // for-comprehension to extract from the two vals
+    val importPerm = for {
+      hasProject <- billingProjects.map (_.exists(_.creationStatus == CreationStatuses.Ready))
+      hasWorkspace <- workspaces.map { ws => ws.exists(_.accessLevel.compare(WorkspaceAccessLevels.Write) >= 0)} // TODO: also need to check canCompute?
+    } yield UserImportPermission(
+      billingProject = hasProject,
+      writableWorkspace = hasWorkspace)
+
+    // respond with a UserImportPermission
+    importPerm map { x=> RequestComplete(StatusCodes.OK, x)}
```
Comment (aednichols) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/UserService.scala]: I wonder if there's a way to roll this extra `map` into the for-comp

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/ImportPermissionApiServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        WorkspaceListResponse(WorkspaceAccessLevels.NoAccess, newWorkspace, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false))
+      ))
+      case "noWorkspaces" => Future.successful(Seq.empty[WorkspaceListResponse])
+      case _ => Future.failed(new FireCloudException("intentional exception for getWorkspaces catchall case"))
+    }
+  }
+
+  // this is hacky, but the only argument to getProjects and getWorkspaces is the access token. Therefore,
+  // we need to encode our test criteria into a string, and we can do so using a delimeter.
+  private def parseTestToken(userInfo: WithAccessToken): (String,String) = {
```
Comment (aednichols) [inline src/test/scala/org/broadinstitute/dsde/firecloud/webservice/ImportPermissionApiServiceSpec.scala]: Nice.

**Review by aednichols — APPROVED** [review]
LGTM 👍

Thinking out loud, no need to change: I do wonder if this would make a good Rawls feature.

Rawls already has a `listProjectsWithCreationStatus(status: CreationStatuses.CreationStatus)` query; we could add a sibling
```
def atLeastOneReadyBillingProject(): ReadWriteAction[Boolean] = {
  for {
    ready <- filter(_.creationStatus === CreationStatuses.Ready.toString).exists
  } yield {
    ready
  }
}
```
and let SQL do its thing efficiently and then return a boolean to orch instead of a BP list. Something similar is probably possible for workspaces.

A side-benefit would be saving adding the new service to Orch

**Comment by davidangb** [comment]
I'd prefer to leave rawls generic, and have orch contain the convenience code for the UI. Totally get your point about efficiency wanting to be as low in the stack as possible, but I lean towards the genericness on this particular choice.

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/ImportPermissionApiServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        WorkspaceListResponse(WorkspaceAccessLevels.Read, publishedRawlsWorkspaceWithAttributes, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false)),
+        WorkspaceListResponse(WorkspaceAccessLevels.NoAccess, newWorkspace, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false))
+      ))
+      case "noWorkspaces" => Future.successful(Seq.empty[WorkspaceListResponse])
+      case _ => Future.failed(new FireCloudException("intentional exception for getWorkspaces catchall case"))
+    }
+  }
+
+  // this is hacky, but the only argument to getProjects and getWorkspaces is the access token. Therefore,
+  // we need to encode our test criteria into a string, and we can do so using a delimeter.
```
Comment (rushtong) [inline src/test/scala/org/broadinstitute/dsde/firecloud/webservice/ImportPermissionApiServiceSpec.scala]: nit, typo: `delimeter` -> `delimiter`

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/ImportPermissionApiServiceSpec.scala (hunk 4 above)
Comment (rushtong) [inline src/test/scala/org/broadinstitute/dsde/firecloud/webservice/ImportPermissionApiServiceSpec.scala]: Yes, this is an interesting approach (👍). The other way I can think of is to create completely separate test classes (each with a custom override) for each of the 4 cases ... but that seems like far more work than necessary.

**Review by rushtong — APPROVED** [review]
👍 Code looks good, tests look comprehensive, and it works for me under different real-world test cases. One minor nit, no need for a re-review.

**Comment by davidangb** [comment]
last few test runs ...

https://fc-jenkins.dsp-techops.broadinstitute.org/job/firecloud-ui-fiab-test-runner/392/consoleFull
```
[info] All tests passed.
[info] Passed: Total 124, Failed 0, Errors 0, Passed 124, Ignored 5
```

https://fc-jenkins.dsp-techops.broadinstitute.org/job/firecloud-ui-fiab-test-runner/398/consoleFull
```
[info] *** 3 TESTS FAILED ***
[error] Failed: Total 124, Failed 3, Errors 0, Passed 121, Ignored 5
[error] Failed tests:
[error] 	org.broadinstitute.dsde.firecloud.test.api.orch.OrchestrationApiSpec
[error] 	org.broadinstitute.dsde.firecloud.test.security.AuthDomainSpec
[error] 	org.broadinstitute.dsde.firecloud.test.metadata.DataSpec

[info] OrchestrationApiSpec:
[info] Orchestration
[info] - should not allow access alteration for arbitrary google roles *** FAILED ***
[info]   java.lang.Exception: Billing project creation did not complete

[info] DataSpec:
[info] Writer and reader should see new columns
[info] - with no defaults or local preferences when analysis run that creates new columns *** FAILED ***
[info]   org.openqa.selenium.TimeoutException: Expected condition failed: waiting for org.broadinstitute.dsde.workbench.service.test.WebBrowserUtil$await$$anon$1@4a3e19bf (tried for 30 second(s) with 500 MILLISECONDS interval)

[info] AuthDomainSpec:
[info] A workspace with an authorization domain
[info]   with one group inside of it
[info]     when the user is inside of the group
[info]       when the workspace is shared with them
[info]       - can be seen and is accessible *** FAILED ***
[info]         org.openqa.selenium.TimeoutException: Expected condition failed: waiting for org.broadinstitute.dsde.workbench.service.test.WebBrowserUtil$await$$anon$1@5b151074 (tried for 30 second(s) with 500 MILLISECONDS interval)
```
https://fc-jenkins.dsp-techops.broadinstitute.org/job/firecloud-ui-fiab-test-runner/401/consoleFull
```
[info] *** 2 TESTS FAILED ***
[error] Failed: Total 127, Failed 2, Errors 0, Passed 125, Ignored 2
[error] Failed tests:
[error] 	org.broadinstitute.dsde.firecloud.test.analysis.MethodConfigSpec

[info] MethodConfigSpec:
[info] - launch a simple workflow *** FAILED ***
[info]   org.openqa.selenium.TimeoutException: Expected condition failed: waiting for org.broadinstitute.dsde.workbench.service.test.WebBrowserUtil$await$$anon$1@1b48d079 (tried for 30 second(s) with 500 MILLISECONDS interval)

[info] - launch modal with no default entities *** FAILED ***
[info]   org.openqa.selenium.TimeoutException: Expected condition failed: waiting for org.broadinstitute.dsde.workbench.service.test.WebBrowserUtil$await$$anon$1@479a48b4 (tried for 30 second(s) with 500 MILLISECONDS interval)
```

**Comment by davidangb** [comment]
( verbal approval from Nicole to merge )

**Pull request merged**